src/codecatalyst/commands.ts

```diff
@@ -135,13 +136,43 @@ function createClientInjector(authProvider: CodeCatalystAuthenticationProvider):
 
         await authProvider.restore()
         const conn = authProvider.activeConnection ?? (await authProvider.promptNotConnected())
-        const client = await createClient(conn)
+        const authenticatedConn = await authenticateConnection(conn, authProvider.auth)
+        const client = await createClient(authenticatedConn)
```

Comment:
This implies that `authenticateConnection` either throws or always succeeds. Should we do this in `createClient` itself then? Maybe this is out of scope of for this PR. It seems like `createClientInjector` might be applicable beyond just "commands.ts".

Reply:
> IMO any client shouldn't need to worry if the connection is valid.

There are many, many edge-cases involved with auth which is why a lot of decisions are delegated. For example, we likely wouldn't want to show a notification for a background task, at least not the same notification. CodeWhisperer is another good example because they might make API requests every time the user types. 

Logic [already exists](https://github.com/aws/aws-toolkit-vscode/blob/965cffb7f317af28369e3a5f7a4f187b91ad9f7b/src/credentials/auth.ts#L684-L694) that checks the local state of _all_ connections before they're used but we are conservative about when notifications are shown to avoid spamming the user.

Another really important thing to keep in mind is that a connection can only really be deemed "valid" _after_ making the desired request. I did not implement any logic to back-propagate exceptions from client requests to `Auth.ts` though this would help a lot with keeping states in-sync.